core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java

```diff
@@ -320,10 +322,12 @@ public int read(ByteBuffer byteBuffer, int off, int len) throws IOException {
     }
     if (mEOF) {
       closeDataReader();
-      Preconditions.checkState(mPos >= mLength,
-          "Block %s is expected to be %s bytes, but only %s bytes are available. "
-              + "Please ensure its metadata is consistent between Alluxio and UFS.",
-          mId, mLength, mPos);
+      if (mPos < mLength) {
```

Comment:
let's make it `if (mPos != mLength)`? WDYT @tcrain @yuzhu

Reply:
Yes, I wonder if it is also possible to remove lines 316-317, because in that case you could still return successfully even if the you haven't yet got an EOF?

Reply:
@JiamingMai see above

Reply:
I guess a set of rules like this would make sense
- Only return -1 when EOF is read
- If EOF read and not the expected length throw exception and do the sync
- If read past the expected length throw exception and do a sync

Reply:
> I guess a set of rules like this would make sense
> 
> * Only return -1 when EOF is read
> * If EOF read and not the expected length throw exception and do the sync
> * If read past the expected length throw exception and do a sync

It will never read past the expected length as it reads the file with the given length received from master.

Reply:
I was suggesting this for two reasons 
- Logically it seems cleaner to only return -1 when an EOF is reached, otherwise you could be in some error condition
- Since we are erroring out when the actual length is shorted then the expected length, should we also not error out when the actual length is longer than the expected length?
- I know that we have no true guarantees that what the reader is reading is consistent, but maybe this is a little better?

Reply:
Discussed with @JiamingMai on this. The current Alluxio stream read logic is we just read until the length. For example the AlluxioFileInStream [here](https://github.com/Alluxio/alluxio/blob/446a255d348485c2d5378042d1d87753a1a3fc24/core/client/fs/src/main/java/alluxio/client/file/AlluxioFileInStream.java#L181) only reads based on the length from file metadata. 

So our conclusion is that it changes many moving parts if we now allow it to read until the EOF(not stopping at length) and check how much we read. So we are going to keep the current `if (mPos < mLength)` logic.

Taking a step back, the current Alluxio cannot handle the problem where the in-UFS file has been replaced with the same or a greater length. Created https://github.com/Alluxio/alluxio/issues/16225 for a future design to address.

Reply:
I don't know exactly the workload that is causing this issue so if this fixes it, and we can track the remaining issues, then I am ok with this.

But I also want to note that, if I understand correctly, in some sense this fix worries me more because we are changing the code to somehow partially silently correct an inconsistency, but allowing other inconsistencies to persist.

For example in some sense it seems that allowing the read to fail continuously to the user is a good thing as then they will realize that they are doing something that is creating inconsistencies, allowing them to either fix their workload or manually perform syncs when necessary. Now with this fix their workload may succeed in more cases where inconsistent results are created at the end.

Reply:
On inconsistency, the read will fail. What this change adds is a forced sync which gives a retry a better chance to succeed. So the user does need to trigger a retry on whatever workflow like submitting a SQL or triggering a pipeline. I hope the user is then able to realize something is wrong and checks the underlying workflow, if he/she triggers a rerun for multiple times.